### PR TITLE
[9] Allow the image paste on nodes in diagram from clipboard

### PR DESCRIPTION
**Pull request by scosta-obeo**
Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/9

### CODE REVIEW

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+     * Default constructor.
+     */
+    public PasteImageAction() {
+        super();
+
+        setId(ActionIds.PASTE_IMAGE);
+        setText(Messages.PasteImageAction_text);
+        setToolTipText(Messages.PasteImageAction_toolTipText);
+        setAccelerator(SWT.CTRL | SWT.ALT | 'V');
+
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:82]: This new action should be documented in the user documentation.
Are you sure about the keyboard shortcut? It does not seem to work on my environment (linux)

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    public void updateState() {
+        setEnabled(canBeEnabled());
+    }
+    
+    /**
+     * {@inheritDoc}
+     * 
+     * @see #org.eclipse.sirius.diagram.ui.tools.internal.actions.layout.PasteImageAction PasteImageAction class
+     *      documentation
+     */
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:107]: You can remove this kind of comments automatically added

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+import org.eclipse.sirius.viewpoint.BasicLabelStyle;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.dnd.Clipboard;
+import org.eclipse.swt.dnd.ImageTransfer;
+import org.eclipse.swt.graphics.ImageData;
+import org.eclipse.swt.graphics.ImageLoader;
+import org.eclipse.swt.widgets.Display;
+
+/**
+ * This action provides past image from the clipboard.
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:49]: duplicated comment

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+        // we check if the clipboard contains image and we have at least one element selected
+        if (imgData != null && !selectedStyles.isEmpty()) {
+            // get the first style to provide the context for file creation and domain for command
+            BasicLabelStyle firstStyle = selectedStyles.stream().findFirst().orElseThrow();
+            TransactionalEditingDomain domain = TransactionUtil.getEditingDomain(firstStyle);
+
+            if (domain != null) {
+                // prepare image file creation
+                String filename = ImageManager.generateName("png"); //$NON-NLS-1$
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:121]: That means the image will always have png extension?

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                ImageManager imageManager = ImageManagerProvider.getImageManager();
+                try (InputStream imageStream = getImageStream(imgData)) {
+                    CreateFileFunc createFileFunc = imageManager.getCreateFileFunc(firstStyle, filename, imageStream);
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:124]: The interest of using the CreateFileFunc is not very obvious to me.
It seems that having the ImageManager which could directly return the Supplier would do the job?

File: plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/api/image/ImageManager.java [hunk 6]
```diff
@@ -36,7 +49,89 @@ public interface ImageManager {
...
+
+    /**
+     * This class store the function that creates the file and its file name. Result of {@link #getCreateFileFunc}
+     * 
+     * @author scosta
+     *
+     */
+    final class CreateFileFunc {
+
+        private Supplier<Optional<CoreException>> createFileFunc;
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/api/image/ImageManager.java:64]: I was wondering if we could not just having the Supplier<Optional<String>> which could return the computed fileName (passed in its constructor maybe) when the get() is called

**Review by florianbarbin — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java (hunk 4 above)
Comment (scosta-obeo) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java:121]: Yes, I always save the image in PNG, see https://github.com/scosta-obeo/sirius-desktop/blob/9766cb963a29a1693c940669fc3f5d947dd91e43/plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java#L162

File: plugins/org.eclipse.sirius.doc/doc/user/diagrams/Diagrams.textile [hunk 7]
```diff
@@ -738,6 +738,12 @@ The result is the same as applying _Paste Format_ and _Paste Style_.
 
 	!images/diagram_editor_paste_format2.png!
 
+h3(#paste_image). Paste image from clipboard
+
+It is also possible to paste an image from the clipboard to a node or container. To do this, copy image in your clipboard, select a node and click on button "Paste image from clipboard" (see picture below). The image is saved as a PNG file in the workspace, in an "images" folder and the selected element has a workspace image style that references this image.
```
Comment (florianbarbin) [inline plugins/org.eclipse.sirius.doc/doc/user/diagrams/Diagrams.textile:743]: To do this, copy image in your clipboard => To do this, copy **the** image in your clipboard, select a node and click on **the** button "Paste image from clipboard"

**Review by florianbarbin — APPROVED** [review]

**Pull request merged**